### PR TITLE
use upstream drain library instead of local modification.

### PR DESCRIPTION
**Pull request by MoserMichael**
Signed-off-by: Michael Moser mmoser@redhat.com

**What this PR does / why we need it:**

Use upstream drain library for draining of pods from node that is sent into maintenance mode. The upstream drain library is k8s.io/kubectl/pkg/drain

**Special notes for your reviewer:**

**Release note:**
```release-note
Use upstream drain library for draining of pods from node that is sent into maintenance mode. The upstream drain library is k8s.io/kubectl/pkg/drain
```

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 1]
```diff
@@ -206,16 +207,25 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
...
-	if err = drainPods(r, node, stop); err != nil {
-		return r.reconcileAndError(instance, err)
+	podsToDelete, errors := r.drainer.GetPodsForDeletion(nodeName)
+
+	if errors != nil && len(errors) != 0 {
+		aerror := utilerrors.NewAggregate(errors)
+		reqLogger.Errorf("Errors while listing pods for deletion: %v", aerror)
+	}
+
+	if len(podsToDelete.Pods()) != 0 {
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: What is the purpose of this new functionality?

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 2]
```diff
@@ -206,16 +207,25 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
 		return r.reconcileAndError(instance, err)
 	}
 
-	if err = runCordonOrUncordon(r, node, true); err != nil {
+	if err = drain.RunCordonOrUncordon(r.drainer, node, true); err != nil {
 		return r.reconcileAndError(instance, err)
 	}
 
 	stop := make(chan struct{})
```
Comment (yanirq) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: revisit this stop channel if its needed , it was passed into the old drainer replaced now with the drain lib.

Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: deleted the channel in commit f4a5380497138abe83d95afea1abd1940f0e0e61

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 1 above)
Comment (yanirq) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: This is done under initMaintenance for status purposes: 
https://github.com/kubevirt/node-maintenance-operator/blob/9fe31971cf36b090fbc6eee31f687c8cb30c3fdb/pkg/controller/nodemaintenance/nodemaintenance_controller.go#L286
And under reconcile and error:
https://github.com/kubevirt/node-maintenance-operator/blob/9fe31971cf36b090fbc6eee31f687c8cb30c3fdb/pkg/controller/nodemaintenance/nodemaintenance_controller.go#L311

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 3]
```diff
@@ -206,16 +207,22 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
...
 
-	reqLogger.Infof("Evict all Pods from Node: %s", nodeName)
-	if err = drainPods(r, node, stop); err != nil {
-		return r.reconcileAndError(instance, err)
+	if errors != nil && len(errors) != 0 {
+		aerror := utilerrors.NewAggregate(errors)
+		reqLogger.Errorf("Errors while listing pods for deletion: %v", aerror)
+	}
+
+	if len(podsToDelete.Pods()) != 0 {
```
Comment (yanirq) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: since you are not reconciling again if `GetPodsForDeletion` returns an error can you end up with podsToDelete being nil ? if so then you will get an unexpected error here with calling Pods() on nil.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 1 above)
Comment (beekhof) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: > we need to list the pods that are to be evicted prior to calling the DeleteOrEvictPods method of the upstream drain library.
> 217: GetPodsForDeletion - the call returns the list of pods to be evicted from the node
> 219-222: report errors if necessary. create a new aggregate error that includes the array of errors returned by GetPodsForDeletion
> 224: check if there is anything to evict prior to calling the evict call.

I understand what the new code is doing, but I'm yet to hear the reason why it is necessary.

**Comment by beekhof** [comment]
Git usage is much improved in this PR

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 4]
```diff
@@ -114,7 +127,10 @@ func initDrainer(r *ReconcileNodeMaintenance, config *rest.Config) error {
 	r.drainer.Client = cs
 	r.drainer.DryRun = false
 
-	return nil
+	r.drainer.Out = writer{klog.Info}
+	r.drainer.ErrOut = writer{klog.Error}
+
+    return nil
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: No need for `OnPodDeletedOrEvicted` ?

Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: OnPodDeletedOrEvicted callback is optional - the drainer only calls it if it has been set (field value is not nil). This differs from the Out fields - it is accessed without a check, so that the Out field is mandatory.  I could set the OnPodDeletedOrEvicted callback to log the event that a pod has been evicted. Do you request to do so?

Comment (beekhof) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Yes please

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 5]
```diff
@@ -86,6 +86,11 @@ func (w writer) Write(p []byte) (n int, err error) {
 	return len(p), nil
 }
 
+func onPodDeletedOrEvicted(pod *corev1.Pod, usingEviction bool) {
+	msg := fmt.Sprintf("pod: %s:%s evicted from node: %s method: %t", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name, pod.Spec.NodeName, usingEviction)
+	klog.Info(msg)
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]:
```suggestion
        verbStr := "Deleted"
        if usingEviction {
            verbStr = "Evicted"
        }
        klog.Info(fmt.Sprintf("%s pod from Node", verbStr),
                  "pod", fmt.Sprintf("%s/%s", pod.Name, pod.Namespace))
```

Comment (beekhof) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: I think I like the original better.
True and false don't clearly relate to Deleted and Evicted

Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: changed the log message and force pushed the change into the last commit - so as not to introduce another commit just for the purpose of changing a printf.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 3 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: changed in a later commit. please review.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 1 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: changed in later commit, please review.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 4 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: has been done in subsequent PR. please review.

File: test/e2e/nodemaintenance_test.go [hunk 6]
```diff
@@ -6,6 +6,8 @@ import (
 	"reflect"
 	"testing"
 	"time"
+    "os"
+    "os/exec"
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: Nit: inconsistent whitespace

File: test/e2e/nodemaintenance_test.go [hunk 7]
```diff
@@ -142,6 +164,7 @@ func nodeMaintenanceTest(t *testing.T, f *framework.Framework, ctx *framework.Te
 		return true, nil
 	}); err != nil {
 		checkFailureStatus(t, f)
+        showDeploymentStatusScript(t)
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: Nit: whitespace. spaces -> tabs for consistency

**Comment by beekhof** [comment]
A couple of whitespace nits and the shared informer can be removed as discussed, but looking promising.

**Comment by beekhof** [comment]
You also need to follow the release note process mentioned in the first comment.
Applies to all PRs of released projects

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 8]
```diff
@@ -153,9 +153,9 @@ type ReconcileHandler interface {
 type ReconcileNodeMaintenance struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
+	client	  client.Client
+	scheme	  *runtime.Scheme
+	drainer	 *drain.Helper
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Apologies for being unclear...
Do not change whitespace in lines that were not otherwise modified by your patch.
The idea is to not introduce additional inconsistencies in _leading_ whitespace only.

Comment (yanirq) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Usually the IDE (using indent plugins) will align those. 
There is also a whitespace check in our make that will fail the build https://github.com/kubevirt/node-maintenance-operator/blob/9fe31971cf36b090fbc6eee31f687c8cb30c3fdb/Makefile#L47   
But that doesn't apply for indents.

**Comment by beekhof** [comment]
Can we clean up the git history here please?
Subsequent commits that fix whitespace, or improve code that you added as part of the PR should be merged into the commit that first introduced the change.

There are commits here that mix fixes and unrelated whitespace changes.  Do not do this.

Also, please address https://github.com/kubevirt/node-maintenance-operator/pull/60#discussion_r393400068

**Comment by MoserMichael** [comment]
> Can we clean up the git history here please?
> Subsequent commits that fix whitespace, or improve code that you added as part of the PR should be merged into the commit that first introduced the change.
> 
> There are commits here that mix fixes and unrelated whitespace changes. Do not do this.
> 
> Also, please address [#60 (comment)](https://github.com/kubevirt/node-maintenance-operator/pull/60#discussion_r393400068)

Right now it seems to be very difficult to work with the CI system - any job submitted takes ages (i see that with the other PR's)

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 8 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: I converted all spaces to tabs using the enforce-tabs-or-spaces.sh scripts from pr #66. Now upon repeated reading that contradicts with your recommendation from conversation #60, I am at a loss what to do now.

**Comment by yanirq** [comment]
You need to run `make gen-k8s` when manipulating CRDs . 
We need to add this to the documentation.  (you will see it as an error and a request in the travis build log)

File: cmd/manager/main.go [hunk 9]
```diff
@@ -25,7 +25,7 @@ import (
 
 // Change below variables to serve metrics on different host or port.
 var (
-	metricsHost       = "0.0.0.0"
+	metricsHost		  = "0.0.0.0"
```
Comment (beekhof) [inline cmd/manager/main.go]: this breaks the alignment of the values

File: cmd/manager/main.go [hunk 10]
```diff
@@ -83,7 +83,7 @@ func main() {
 
 	// Create a new Cmd to provide shared dependencies and start components
 	mgr, err := manager.New(cfg, manager.Options{
-		Namespace:          namespace,
+		Namespace:			namespace,
```
Comment (beekhof) [inline cmd/manager/main.go]: this breaks the alignment of the values

File: pkg/apis/kubevirt/v1alpha1/nodemaintenance_types.go [hunk 11]
```diff
@@ -33,10 +33,10 @@ type NodeMaintenanceSpec struct {
 // kubebuilder:subresource:status
 // +k8s:openapi-gen=true
 type NodeMaintenance struct {
-	metav1.TypeMeta   `json:",inline"`
+	metav1.TypeMeta	  `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   NodeMaintenanceSpec   `json:"spec,omitempty"`
+	Spec   NodeMaintenanceSpec	 `json:"spec,omitempty"`
```
Comment (beekhof) [inline pkg/apis/kubevirt/v1alpha1/nodemaintenance_types.go]: this breaks the alignment of the json spec

File: pkg/apis/kubevirt/v1alpha1/nodemaintenance_types.go [hunk 12]
```diff
@@ -46,7 +46,7 @@ type NodeMaintenance struct {
 type NodeMaintenanceList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []NodeMaintenance `json:"items"`
+	Items			[]NodeMaintenance `json:"items"`
```
Comment (beekhof) [inline pkg/apis/kubevirt/v1alpha1/nodemaintenance_types.go]: this breaks the alignment of the types

File: pkg/apis/kubevirt/v1alpha1/register.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// NOTE: Boilerplate only.  Ignore this file.
+// NOTE: Boilerplate only.	Ignore this file.
```
Comment (beekhof) [inline pkg/apis/kubevirt/v1alpha1/register.go]: unecessary

File: pkg/apis/kubevirt/v1alpha1/zz_generated.openapi.go [hunk 14]
```diff
@@ -26,15 +26,15 @@ func schema_pkg_apis_kubevirt_v1alpha1_NodeMaintenance(ref common.ReferenceCallb
...
+							Format:		 "",
 						},
 					},
 					"apiVersion": {
 						SchemaProps: spec.SchemaProps{
 							Description: "APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources",
-							Type:        []string{"string"},
-							Format:      "",
+							Type:		 []string{"string"},
+							Format:		 "",
```
Comment (beekhof) [inline pkg/apis/kubevirt/v1alpha1/zz_generated.openapi.go]: Dont modify generated files

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 15]
```diff
@@ -154,9 +153,9 @@ type ReconcileHandler interface {
...
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Still wrong, types are no longer aligned

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 16]
```diff
@@ -237,18 +236,9 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
...
-	}
-
-	if len(podsToDelete.Pods()) != 0 {
-		reqLogger.Infof("Evict all Pods from Node: %s", nodeName)
-		if err = r.drainer.DeleteOrEvictPods(podsToDelete.Pods()); err != nil {
-			return r.reconcileAndError(instance, err)
-		}
+	err = drain.RunNodeDrain(r.drainer, nodeName)
+	if err != nil {
+		return r.reconcileAndError(instance, err)
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: please do not mix functional and whitespace changes in the same commit

File: test/e2e/nodemaintenance_test.go [hunk 17]
```diff
@@ -24,12 +22,12 @@ import (
...
-	timeout              = time.Second * 120
+	retryInterval		= time.Second * 5
+	timeout			  = time.Second * 120
 	cleanupRetryInterval = time.Second * 1
-	cleanupTimeout       = time.Second * 5
-	testDeployment       = "testdeployment"
-	podLabel             = map[string]string{"test": "drain"}
+	cleanupTimeout	   = time.Second * 5
+	testDeployment	   = "testdeployment"
+	podLabel			 = map[string]string{"test": "drain"}
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: test/e2e/nodemaintenance_test.go [hunk 18]
```diff
@@ -55,7 +53,7 @@ func showDeploymentStatusScript(t *testing.T) {
 func TestNodeMainenance(t *testing.T) {
 	nodeMainenanceList := &operator.NodeMaintenanceList{
 		TypeMeta: metav1.TypeMeta{
-			Kind:       "NodeMaintenance",
+			Kind:	   "NodeMaintenance",
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: test/e2e/nodemaintenance_test.go [hunk 19]
```diff
@@ -116,15 +114,15 @@ func nodeMaintenanceTest(t *testing.T, f *framework.Framework, ctx *framework.Te
 	// Create node maintenance custom resource
 	nodeMaintenance := &operator.NodeMaintenance{
 		TypeMeta: metav1.TypeMeta{
-			Kind:       "NodeMaintenance",
+			Kind:	   "NodeMaintenance",
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: test/e2e/nodemaintenance_test.go [hunk 20]
```diff
@@ -260,10 +258,10 @@ func createSimpleDeployment(t *testing.T, f *framework.Framework, ctx *framework
 	dep := &appsv1.Deployment{
 		TypeMeta: metav1.TypeMeta{
 			APIVersion: "apps/v1",
-			Kind:       "Deployment",
+			Kind:	   "Deployment",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      testDeployment,
+			Name:	  testDeployment,
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: test/e2e/nodemaintenance_test.go [hunk 21]
```diff
@@ -274,7 +272,7 @@ func createSimpleDeployment(t *testing.T, f *framework.Framework, ctx *framework
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
 					Namespace: namespace,
-					Labels:    podLabel,
+					Labels:	podLabel,
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: test/e2e/nodemaintenance_test.go [hunk 22]
```diff
@@ -294,10 +292,10 @@ func createSimpleDeployment(t *testing.T, f *framework.Framework, ctx *framework
 						},
 					},
 					Containers: []corev1.Container{{
-						Image:   "busybox",
-						Name:    "testpodbusybox",
+						Image:	 "busybox",
+						Name:	"testpodbusybox",
 						Command: []string{"/bin/sh"},
-						Args:    []string{"-c", "while true; do echo hello; sleep 10;done"},
+						Args:	[]string{"-c", "while true; do echo hello; sleep 10;done"},
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: this breaks the alignment of values

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 23]
```diff
@@ -127,9 +153,9 @@ type ReconcileHandler interface {
 type ReconcileNodeMaintenance struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
+	client	  client.Client
+	scheme	  *runtime.Scheme
+	drainer	 *drain.Helper
 	podInformer cache.SharedInformer
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: This still doesn't look right.

File: test/e2e/nodemaintenance_test.go (hunk 6 above)
Comment (beekhof) [reply test/e2e/nodemaintenance_test.go]: +1

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 23 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: fixed that.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 24]
```diff
@@ -206,15 +232,12 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
...
 		return r.reconcileAndError(instance, err)
 	}
 
-	stop := make(chan struct{})
-	defer close(stop)
-
-	reqLogger.Infof("Evict all Pods from Node: %s", nodeName)
-	if err = drainPods(r, node, stop); err != nil {
+	err = drain.RunNodeDrain(r.drainer, nodeName)
+	if err != nil {
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go]: Lets follow the existing convention...
```suggestion
	if err = drain.RunNodeDrain(r.drainer, nodeName); err != nil {
```

File: test/e2e/nodemaintenance_test.go [hunk 25]
```diff
@@ -29,6 +31,26 @@ var (
...
+	if err != nil {
+		t.Logf("can't get cluster status: %v\n", err)
+	}
+
+	if output != nil {
+		soutput := string(output)
+		t.Logf("cluster status: %s", soutput)
+	}
+}
+
```
Comment (beekhof) [inline test/e2e/nodemaintenance_test.go]: Useful, but doesn't belong in this PR

Comment (MoserMichael) [reply test/e2e/nodemaintenance_test.go]: removed from this PR, added new PR #69 with this change.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go (hunk 24 above)
Comment (MoserMichael) [reply pkg/controller/nodemaintenance/nodemaintenance_controller.go]: force pushed as part of the last commit.

File: pkg/controller/nodemaintenance/nodemaintenance_controller.go [hunk 26]
```diff
@@ -206,15 +232,11 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
...
 
-	if err = runCordonOrUncordon(r, node, true); err != nil {
+	if err = drain.RunCordonOrUncordon(r.drainer, node, true); err != nil {
 		return r.reconcileAndError(instance, err)
 	}
 
-	stop := make(chan struct{})
-	defer close(stop)
-
-	reqLogger.Infof("Evict all Pods from Node: %s", nodeName)
```
Comment (beekhof) [inline pkg/controller/nodemaintenance/nodemaintenance_controller.go:216]: Any reason for removing this log message?

**Comment by beekhof** [comment]
I can't find the comment where @yanirq asked for it, but we still need the new dependancy to be listed in Gopkg.toml
Then I will approve.

**Comment by beekhof** [comment]
> I can't find the comment where @yanirq asked for it, but we still need the new dependancy to be listed in Gopkg.toml

Ah, it was in a different PR but is equally relevant here
    https://github.com/kubevirt/node-maintenance-operator/pull/64/files#r401697177

File: Gopkg.toml [hunk 27]
```diff
@@ -18,6 +18,9 @@ required = [
   source = "https://github.com/fsnotify/fsnotify/archive/v1.4.7.tar.gz"
   name = "gopkg.in/fsnotify.v1"
 
+[[override]]
+  name = "k8s.io/kubectl/pkg/drain"
+  version = "kubernetes-1.13.1"
```
Comment (beekhof) [inline Gopkg.toml:23]: There is no version of the kubectl project called `kubernetes-1.13.1` and even if there were, thats not the version we have stored in the vendor directory.

Comment (MoserMichael) [reply Gopkg.toml]: fixed.
